improved and updated spec

erezmus opened the pull request:
Related to #14,

Updated the storage rest and graphql spec

erezmus committed:
improved and updated spec

erezmus commented on storage/docs/graphql.md:
```diff
@@ -1,45 +1,98 @@
+    "updated": "2019-08-19T14:40:04.123456",
+    "size": 12345,
+    "publicLink": "http://storage-url/path/to/file>",
+    "signedLink": "http://storage-url/path/to/file?token=abcdef>",
```
Is the signedLink always returned or do you need to specify it as part of the request?

diversemix replied:
you need to specify it - you don't want it auto generated if you don't specify it.

erezmus commented on storage/docs/rest.md:
````diff
@@ -1,69 +1,83 @@
 ```
-PUT /files/namespace/directory/file.ext
+PUT /files/libero/directory/file.pdf
+
````
Do we need to send the Content-Type?

diversemix replied:
What happened to `namespace` we need that for multi tenanting ... its essentially  a compound of `org` and `journal`

erezmus replied:
see a few lines above:

> To store a file with a namespace of ```libero``` and a key of ```directory/file.pdf```, the following request should be used.

Was thinking that having a more concrete example might be better for illustration

diversemix replied:
Yes.. I understand now ... maybe the abstract one should have been written like:

```
PUT /files/<namespace>/<directory>/<file.ext>
```

erezmus commented on storage/docs/rest.md:
```diff
@@ -1,69 +1,83 @@
-Link: "<http://sharedlink>;rel=????"
-Link: "<http://publiclink>;rel=????"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+Link: "<http://storage-url/path/to/file>;rel=public"
```
What if underlying storage is not public? should it return no link?

petereast replied:
Probably

diversemix commented on storage/docs/graphql.md:
```diff
@@ -1,45 +1,98 @@
+}
+
 type FileMeta {
-  id: ID!
```
Why have we lost the ID ?

erezmus replied:
replaced by 'key'

diversemix replied:
Why don't we have both ... the `key` is not the same thing as `ID` ... at least not from an S3 perspective

erezmus committed:
- Readded id
- Changed signedLink/publicLink mechanism
- Fixed graphql queries and mutations

thewilkybarkid commented on storage/docs/rest.md:
````diff
@@ -1,69 +1,117 @@
+then a 412 (Precondition Failed) should be returned.
+
+The ```Libero-file-id``` value is a reference to the storage entity's internal id for a file. When uploading to the
+same URL, a new value for ```Libero-file-id``` should be returned.
````
This is confusing. Are files meant to be modifiable? If so, this looks like an Etag?

erezmus replied:
@diversemix is the id redundant if we have etags?

diversemix replied:
I would say not - Etags and ID have two different purposes and are the responsibilities of two different systems. I'd say we are going into too much detail at this stage - remember this is a demo.

thewilkybarkid replied:
So don't include this?

diversemix replied:
I'd say this may prove difficult to define and implement by the end of the day - so seems like a good feature to loose for expediency - My take is that its an app-specific field that should be returned in the response - but I don't think we have time to discuss this right now.

thewilkybarkid commented on storage/docs/rest.md:
````diff
@@ -1,69 +1,117 @@
 
-## Get a file
+When uploading for the first time, the ```If-Match``` header should be set to '*'. If the ```If-Match``` fails,
+then a 412 (Precondition Failed) should be returned.
````
Could make it required and return a [`428 Precondition Required`](https://tools.ietf.org/html/rfc6585#section-3) if missing?

erezmus replied:
isn't 428 for when If-Match header is missing? maybe i understood it wrong...

thewilkybarkid replied:
Yeah, that's what I meant. (Safer writes, but depends if they should be modifiable; refs https://github.com/libero/shared-services-experiment/pull/37#discussion_r317096939).

diversemix replied:
Do we need this header for the demo?

thewilkybarkid commented on storage/docs/rest.md:
```diff
@@ -1,69 +1,117 @@
 Response:
 Code: 201
-Link: "<http://url/foo/bar.pdf>;rel=signed"
+Link: "<http://storage-url/path/to/file>;rel=public"
```
`public` isn't a registered type (https://www.iana.org/assignments/link-relations/link-relations.xhtml) so it will need to be a URI (`http://libero.pub/public` or something; see https://tools.ietf.org/html/rfc8288#section-2.1.2).

erezmus replied:
The link would actually be a direct link to the underlying storage url. So for s3 it would be something like:

https://mybucket.s3.amazonaws.com/directory/file.pdf

can we use the "original" type for example?

thewilkybarkid replied:
`original` has a different meaning (original state of a resource).

Don't there there is one that describes this case (which is a publicly accessible version, so could be a CDN URI rather than a bucket directly).

thewilkybarkid commented on storage/docs/rest.md:
````diff
@@ -1,69 +1,117 @@
+The Link value should be a signed url that can subsquently be used to access the file directly from
+the private storage entity.
+
+Note that if the ```Libero-url-type``` header is not sent then no Link header should be returned in the response.
````
Definitely needs a `Vary`; pretty custom though. If it's _really_ needed (ie what's the problem with always returning it if you have access?), then a [`Prefer` header](https://tools.ietf.org/html/rfc7240) is preferable (😎).

erezmus replied:
So the problem is that you don't want to have to query, say s3 to get the signed url. We had this issue in current xpub where we were doing that on autosave every 5 seconds, which made no sense (iirc) so you had to ask for it specifically.

Would the header be:
```
Prefer: signed-url
```
?

thewilkybarkid replied:
Yeah, or something like `Prefer: public-link=signed` if these a variants.

Response needs a `Preference-Applied` and `Vary`.

erezmus committed:
tweaked response headers and rel values

thewilkybarkid commented on storage/docs/rest.md:
```diff
@@ -1,69 +1,125 @@
 Response:
 Code: 201
-Link: "<http://url/foo/bar.pdf>;rel=signed"
+Link: "<http://storage-url/path/to/file>;rel=public"
```
```suggestion
Link: "<http://storage-url/path/to/file>;rel=https://libero.pub/public"
```

erezmus replied:
thanks, done.

erezmus commented on storage/docs/graphql.md:
```diff
@@ -1,45 +1,145 @@
 }
 
+input FileMetaInput {
+  key: String! # link from the service
```
should we not be sending the Content-Type like in rest?

erezmus committed:
added http to code blocks and content length header in rest upload

erezmus committed:
fixed rel for Link header

erezmus committed:
removed id

giorgiosironi commented on storage/docs/rest.md:
````diff
@@ -1,69 +1,118 @@
+
+```http
+HEAD /files/libero/directory/file.pdf
+Prefer: link=signed
````
shouldn't this be `link=https://libero.pub/signed`?

thewilkybarkid replied:
Could be for consistency (might have to be quoted?), but nothing more.
